src/ConnectedMode/QualityProfiles/OutOfDateQualityProfileFinder.cs

```diff
@@ -0,0 +1,93 @@
+ */
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.Binding;
+using SonarQube.Client;
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.ConnectedMode.QualityProfiles
+{
+
+    internal interface IOutOfDateQualityProfileFinder
+    {
+        /// <summary>
+        /// Gives the list of outdated quality profiles based on the existing ones from <see cref="BoundSonarQubeProject.Profiles"/>
+        /// </summary>
+        Task<IEnumerable<(Language language, SonarQubeQualityProfile qualityProfile)>> GetAsync(
+            BoundSonarQubeProject sonarQubeProject,
+            CancellationToken cancellationToken);
+    }
+
+    [Export(typeof(IOutOfDateQualityProfileFinder))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class OutOfDateQualityProfileFinder : IOutOfDateQualityProfileFinder
+    {
+        private readonly ISonarQubeService sonarQubeService;
+
+        [ImportingConstructor]
+        public OutOfDateQualityProfileFinder(ISonarQubeService sonarQubeService)
+        {
+            this.sonarQubeService = sonarQubeService;
+        }
+
+        public async Task<IEnumerable<(Language language, SonarQubeQualityProfile qualityProfile)>> GetAsync(
+            BoundSonarQubeProject sonarQubeProject,
+            CancellationToken cancellationToken)
+        {
+            var sonarQubeQualityProfiles =
+                await sonarQubeService.GetAllQualityProfilesAsync(sonarQubeProject.ProjectKey,
+                    sonarQubeProject.Organization.Key,
+                    cancellationToken);
+            return GetChangedQualityProfiles(sonarQubeProject, sonarQubeQualityProfiles);
+        }
+
+        private static IEnumerable<(Language language, SonarQubeQualityProfile qualityProfile)> GetChangedQualityProfiles(
+            BoundSonarQubeProject sonarQubeProject,
+            IEnumerable<SonarQubeQualityProfile> sonarQubeQualityProfiles)
+        {
+            foreach (var serverQualityProfile in sonarQubeQualityProfiles)
+            {
+                var language = Language.GetLanguageFromLanguageKey(serverQualityProfile.Language);
+                
+                if (language == default)
+                {
+                    continue;
+                }
+                
+                // if we know the language, it's in the dictionary
+                Debug.Assert(sonarQubeProject.Profiles.ContainsKey(language));
+
+                var localQualityProfile = sonarQubeProject.Profiles[language];
+                
+                if (!serverQualityProfile.Key.Equals(localQualityProfile.ProfileKey) 
+                    || serverQualityProfile.TimeStamp > localQualityProfile.ProfileTimestamp)
+                {
+                    yield return (language, serverQualityProfile);
```

Comment:
What's the reason for using `yield` here? We've already done the web call which is the expensive part of the operation so we're not really saving anything and are just adding unnecessary overhead at runtime.